---
title: "BCB420-A1"
output: html_notebook
---
## Packages
load all the packages required
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")
if(!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")
if(!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

library("BiocManager")
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("knitr")

if(!file.exists('GEOmetadb.sqlite')) getSQLiteFile()
```
## Data selection
create a sql searches in the GEOmetadb.sqlite, looking for dataset with submission date after 2015, associated 
with cancer, homo sapians, and high-throughput sequencing.
```{r}
sql <- paste("SELECT DISTINCT gse.title,gse.gse, gpl.title,",
             " gse.submission_date,",
             " gse.supplementary_file",
             "FROM",
             "  gse JOIN gse_gpl ON gse_gpl.gse=gse.gse",
             "  JOIN gpl ON gse_gpl.gpl=gpl.gpl",
             "WHERE",
             "  gse.submission_date > '2015-01-01' AND",
             "  gse.title LIKE '%cancer%' AND", 
             "  gpl.organism LIKE '%Homo sapiens%' AND",
             "  gpl.technology LIKE '%high-throughput sequencing%' ",
             "  ORDER BY gse.submission_date DESC",sep=" ")
con <- dbConnect(SQLite(),'GEOmetadb.sqlite')
rs <- dbGetQuery(con,sql)
unlist(lapply(rs$supplementary_file,
              FUN = function(x){x <- unlist(strsplit(x,";")) ;
              x <- x[grep(x,pattern="txt",ignore.case=TRUE)];
                tail(unlist(strsplit(x,"/")),n=1)})) [1:30]
rs <- dbGetQuery(con,sql)
counts_files <- rs$supplementary_file[grep(rs$supplementary_file,
                                           pattern = "count", ignore.case = TRUE)]

```

#The datasets choosed is GSE162004, which studied how YTHDF2 inhibits gastric cancer cell growth by regulating FOXC2 signaling pathway.
```{r}
sfiles <- getGEOSuppFiles('GSE162004')
fnames <- rownames(sfiles)

```

#load the counts data into R and get its header.
```{r}
data = read.delim(fnames[1],header=TRUE, check.names = FALSE, sep = "\t")
head(data)
kable(head(data), type="html")
```



##Data cleaning 

#What is the dimension of our data and the colomn names of the data?
```{r}
dim(data)
colnames(data)
```
#According to the edgeR protocol, filter weakly expressed and noninformative data. There are 6 samples and therefore the threshold is set to 6.
```{r}
cpms = cpm(data[, 2:7])
rownames(cpms) <- data[,1]

keep = rowSums(cpms >1) >=6
data_filtered = data[keep,]
```
#now check the dimension of the filtered dataset.
```{r}
dim(data_filtered)
```
#The dataset is decreased to 14526 from 53465 after all these cleaning steps.

#Box plot
```{r}
data2plot <- log2(cpm(data_filtered[, 2: ncol(data_filtered)]))
boxplot(data2plot, 
        xlab = "Samples", 
        ylab = "log2 CPM", 
        las = 2, 
        cex = 0.5, 
        cex.lab = 0.5,
        cex.axis = 0.5, 
        main = "RNASeq Samples")
abline(h = median(apply(data2plot,2,median)),
       col = "green",lwd = 0.6, lty = "dashed")
```
#Density Plot
```{r}
counts_density <- 
  apply(log2(cpm(data_filtered[, 2:length(colnames(data_filtered))])),
        2, density)
 #calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
 cols <- rainbow(length(counts_density))
 ltys <- rep(1, length(counts_density))
 
 #plot the first density plot to initialize the plot
 plot(counts_density[[1]], 
      xlim=xlim, 
      ylim=ylim, 
      type="n", 
      ylab="Smoothing density of log2-CPM", 
      main="Density Plot", 
      cex.lab = 0.8)
 
 #plot each line
 for (i in 1:length(counts_density)) {
   lines(counts_density[[i]], 
         col = cols[i], 
         lty = ltys[i])
 }
   
 #create legend
 legend("topright", colnames(data2plot), 
 col=cols, lty=ltys, cex=0.75, 
 border ="blue", text.col = "green4", 
 merge = TRUE, bg = "gray90")
```

#check duplicated genes
```{r}
summarized_gene_counts <- sort(table(data$gene_id),decreasing = TRUE)
kable(summarized_gene_counts[which(summarized_gene_counts > 1)[1:10]], format = 'html')
```
There is no duplicate genes

#Grouping the data
```{r}
samples <- data.frame(lapply(colnames(data)[2:7], FUN=function(x){unlist(strsplit(x, split = "_"))[c(1)]}))
colnames(samples) <- colnames(data)[2:7]
rownames(samples) <- c("condition")
samples <- data.frame(t(samples))
samples
```

## Mapping
#connect to the mart and get the dataset of homosapiens
```{r}
httr::set_config(httr::config(ssl_verifypeer = FALSE)) # sometimes code bellow runs error, this line solves it
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl)
```

#convert the human Ensembl gene ids to HGNC symbols
```{r}
conversion_stash <- "./recq1_id_conversion.rds"
if (file.exists(conversion_stash)) {
  recq1_id_conversion <- readRDS(conversion_stash)
} else {
  recq1_id_conversion <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = data_filtered$gene_id,
                               mart = ensembl)
  saveRDS(recq1_id_conversion, conversion_stash)
}
recq1_id_conversion
```
##Normalization
#creating our DGEList objects to be used by edgeR
```{r}
filter_data_matrix <- as.matrix(data_filtered[,2:7])
rownames(filter_data_matrix) <- data_filtered$gene_id
d = DGEList(counts=filter_data_matrix,group=samples$condition)
```

#Calculate the normalization factors
```{r}
d = calcNormFactors(d)

normalized_counts <- cpm(d)
```

#Inspect the sample separation using a multidimenstional scaling plot.
```{r}
plotMDS(d, labels=rownames(samples),
        col = c("darkgreen","blue")[factor(samples$condition)])
```
##Interpretation
1. What are the control and test conditions of the dataset?
Control condition is the YTHDF2 expression levels in public databses.
test condition i the YTHDF2 expression level in gastric cancer patients samples.

2. Why is the dataset of interest to you?
Gastric cancer is one of the most common malignancies in the world and this dataset study how YTHDF2 inhibits it by regulating the FOXC2 signalling pathway. 

3. Were there expression values that were not unique for specific genes? How did you handle these?
There is no duplicated genes.

4. Were there expression values that could not be mapped to current HUGO symbols?
All the expression values cannot be mapped to current HUGO symbols and there might exist an error 
in the coding, tried but did not come up with a solution. 

5. How many outliers were removed?
no outliers were removed

6. How did you handle replicates?
There is no replicated in my dataset.

7. What is the final coverage of your dataset?
14526 out of 53465


##reference
Lecture modules:https://q.utoronto.ca/courses/248455/files/19147947?module_item_id=3463364

Shen, X., Zhao, K., Xu, L., Cheng, G., Zhu, J., Gan, L., Wu, Y., & Zhuang, Z. (2021). YTHDF2 Inhibits Gastric Cancer Cell Growth by Regulating FOXC2 Signaling Pathway. Frontiers in genetics, 11, 592042. https://doi.org/10.3389/fgene.2020.592042





























































































This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
